File: solutions/typescript/problems/[0141_Easy] Linked List Cycle.ts
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     val: number
 *     next: ListNode | null
 *     constructor(val?: number, next?: ListNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.next = (next===undefined ? null : next)
 *     }
 * }
 */

// Floyd's Cycle Finding Algorithm
function hasCycle(head: ListNode | null): boolean {
    let pointer1 = head || null;
    let pointer2 = head?.next || null;

    while (pointer1 && pointer2) {
        if (pointer1 === pointer2) return true;
        pointer1 = pointer1?.next || null;
        pointer2 = pointer2?.next?.next || null;
    }
    return false;
}

File: solutions/c++/problems/[0141_Easy] Linked List Cycle.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    bool hasCycle(ListNode *head) {
        unordered_map<ListNode*, bool> mp;
        while(head) {
            if (mp[head]) return true;
            mp[head] = true;
            head = head->next;
        }
        return false;
    }
};